### PR TITLE
Manage default printer for the win backend

### PR DESCRIPTION
**Pull request by Pollux42**
Currently, when printer are redirected, the declared default printer is always the first redirected printer.
With this PR, the declared default printer is now the local default printer.

### CODE REVIEW

**Comment by mfleisz** [comment]
@freerdp-bot test

File: channels/printer/client/win/printer_win.c [hunk 1]
```diff
@@ -432,6 +440,15 @@ rdpPrinterDriver* win_freerdp_printer_client_subsystem_entry(void)
 		win_driver->driver.ReleaseRef = printer_win_release_ref_driver;
 
 		win_driver->id_sequence = 1;
+		win_driver->defaultPrinter = NULL;
+
+		GetDefaultPrinter(NULL, &size);
+		if (size)
+		{
+			win_driver->defaultPrinter = (LPWSTR)calloc(size, sizeof(WCHAR));
+			if (!GetDefaultPrinter(win_driver->defaultPrinter, &size))
```
Comment (hardening) [inline channels/printer/client/win/printer_win.c:449]: missing `calloc` return check

**Comment by hardening** [comment]
Question: by setting the default printer at the driver level doesn't it mean that if the default printer is changed by any mean, then we'll continue to use the printer that was set as default at the beginning of the program ?

File: channels/printer/client/win/printer_win.c [hunk 2]
```diff
@@ -408,6 +408,12 @@ static void printer_win_release_ref_driver(rdpPrinterDriver* driver)
 	rdpWinPrinterDriver* win = (rdpWinPrinterDriver*)driver;
 	if (win->references <= 1)
 	{
+		if (win->defaultPrinter)
```
Comment (hardening) [inline channels/printer/client/win/printer_win.c:411]: no need to check for `win->defaultPrinter != NULL` as `free(NULL)` is safe (and does the test internally)

File: channels/printer/client/win/printer_win.c (hunk 1 above)
Comment (Pollux42) [reply channels/printer/client/win/printer_win.c]: Done

File: channels/printer/client/win/printer_win.c (hunk 2 above)
Comment (Pollux42) [reply channels/printer/client/win/printer_win.c]: Done

**Comment by Pollux42** [comment]
Yes, the default printer name is set at the beginning of the program and cannot be changed after.

**Comment by Pollux42** [comment]
I have just realized that the cups backend had the same issue, so I did the same code change for it.

File: channels/printer/client/win/printer_win.c [hunk 3]
```diff
@@ -432,6 +435,19 @@ rdpPrinterDriver* win_freerdp_printer_client_subsystem_entry(void)
...
 
 		win_driver->id_sequence = 1;
+		win_driver->defaultPrinter = NULL;
+
+		GetDefaultPrinter(NULL, &size);
+		if (size)
+		{
+			win_driver->defaultPrinter = (LPWSTR)calloc(size, sizeof(WCHAR));
+
+			if (win_driver->defaultPrinter)
```
Comment (hardening) [inline channels/printer/client/win/printer_win.c:445]: If we can't allocate memory for the default printer's name, I'd rather exit with an error and log a message, than continuing like if everything is ok..

Comment (Pollux42) [reply channels/printer/client/win/printer_win.c:445]: So what about:

		win_driver->defaultPrinter = NULL;

		GetDefaultPrinter(NULL, &size);
		if (size)
		{
			win_driver->defaultPrinter = (LPWSTR)calloc(size, sizeof(WCHAR));

			if (!win_driver->defaultPrinter)
			{
				win_driver->driver.ReleaseRef(&win_driver->driver);
				return NULL;
			}

			if (!GetDefaultPrinter(win_driver->defaultPrinter, &size))
				win_driver->defaultPrinter[0] = "\0";
		}

(And I did the same kind of error message as for if the win_driver allocation fails ;-) )

Comment (Pollux42) [reply channels/printer/client/win/printer_win.c:445]: Or directly `free(win_driver)` instead of `ReleaseRef`?

Comment (hardening) [reply channels/printer/client/win/printer_win.c:445]: What ever works, probably that the direct free is better as this pointer can't have any other references

Comment (Pollux42) [reply channels/printer/client/win/printer_win.c]: Done

**Comment by hardening** [comment]
@freerdp-bot test

**Comment by hardening** [comment]
@Pollux42 run `clang-format` on your changes to pass the code quality checker, and I'll merge it

**Comment by hardening** [comment]
@freerdp-bot can you check and test again please ?

**Pull request merged**